actors/builtin/market/market_actor.go

```diff
@@ -427,78 +411,139 @@ func (a Actor) OnMinerSectorsTerminate(rt Runtime, params *OnMinerSectorsTermina
-	Deals []abi.DealID // TODO: RLE
-}
+func (a Actor) CronTick(rt Runtime, params *adt.EmptyValue) *adt.EmptyValue {
+	rt.ValidateImmediateCallerIs(builtin.CronActorAddr)
+	amountSlashed := big.Zero()
+
+	var timedOutVerifiedDeals []*DealProposal
 
-func (a Actor) HandleExpiredDeals(rt Runtime, params *HandleExpiredDealsParams) *adt.EmptyValue {
-	rt.ValidateImmediateCallerType(builtin.CallerTypesSignable...)
-	var slashed abi.TokenAmount
 	var st State
 	rt.State().Transaction(&st, func() interface{} {
-		slashed = st.updatePendingDealStates(rt, params.Deals, rt.CurrEpoch())
-		return nil
-	})
+		dbe, err := AsSetMultimap(adt.AsStore(rt), st.DealOpsByEpoch)
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "failed to load deal opts set: %s", err)
+		}
 
-	// TODO: award some small portion of slashed to caller as incentive
+		updatesNeeded := make(map[abi.ChainEpoch][]abi.DealID)
 
-	_, code := rt.Send(builtin.BurntFundsActorAddr, builtin.MethodSend, nil, slashed)
-	builtin.RequireSuccess(rt, code, "failed to burn funds")
-	return nil
-}
+		states, err := AsDealStateArray(adt.AsStore(rt), st.States)
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "get state state: %v", err)
+		}
 
-type HandleInitTimeoutDealsParams struct {
-	Deals []abi.DealID // TODO: RLE
-}
+		et, err := adt.AsBalanceTable(adt.AsStore(rt), st.EscrowTable)
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "loading escrow table: %s", err)
+		}
 
-func (a Actor) HandleInitTimeoutDeals(rt Runtime, params *HandleInitTimeoutDealsParams) *adt.EmptyValue {
-	rt.ValidateImmediateCallerType(builtin.CallerTypesSignable...)
-	var st State
-	var verifiedDeals []*DealProposal
-	slashedAmount := rt.State().Transaction(&st, func() interface{} {
-		slashed := abi.NewTokenAmount(0)
-		for _, dealID := range params.Deals {
-			deal := st.mustGetDeal(rt, dealID)
-			state := st.mustGetDealState(rt, dealID)
-
-			// Deal has not been activated.
-			if state.SectorStartEpoch == epochUndefined {
-				// Now is after StartEpoch when the Deal should have been activated, hence clean up.
-				if rt.CurrEpoch() > deal.StartEpoch {
-					// Store VerifiedDeal to restore bytes for VerifiedClient.
-					if deal.VerifiedDeal {
-						verifiedDeals = append(verifiedDeals, deal)
+		lt, err := adt.AsBalanceTable(adt.AsStore(rt), st.LockedTable)
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "loading locked balance table: %s", err)
+		}
+
+		for i := st.LastCron; i <= rt.CurrEpoch(); i++ {
+			if err := dbe.ForEach(i, func(dealID abi.DealID) error {
+				state, found, err := states.Get(dealID)
+				if err != nil {
+					rt.Abortf(exitcode.ErrIllegalState, "failed to get deal: %d", dealID)
+				}
+
+				if !found {
+					return nil
+				}
+
+				deal := st.mustGetDeal(rt, dealID)
+
+				if state.SectorStartEpoch == epochUndefined {
+					// Not yet appeared in proven sector; check for timeout.
+					if rt.CurrEpoch() > deal.StartEpoch {
```

Comment:
The expiry check is scheduled for `deal.StartEpoch` exactly, and cron runs at the end of the epoch, so I think this should be `>=`